node/src/reactor/joiner/memory_metrics.rs

```diff
@@ -0,0 +1,275 @@
+        let address_gossiper = reactor.address_gossiper.estimate_heap_size() as i64;
+        let config = reactor.config.estimate_heap_size() as i64;
+        let chainspec_loader = reactor.chainspec_loader.estimate_heap_size() as i64;
+        let storage = reactor.storage.estimate_heap_size() as i64;
+        let contract_runtime = reactor.contract_runtime.estimate_heap_size() as i64;
+        let linear_chain_fetcher = reactor.linear_chain_fetcher.estimate_heap_size() as i64;
+        let linear_chain_sync = reactor.linear_chain_sync.estimate_heap_size() as i64;
+        let block_validator = reactor.block_validator.estimate_heap_size() as i64;
+        let deploy_fetcher = reactor.deploy_fetcher.estimate_heap_size() as i64;
+        let block_executor = reactor.block_executor.estimate_heap_size() as i64;
+        let consensus = reactor.consensus.estimate_heap_size() as i64;
+
+        let total = metrics
+            + network
+            + small_network
+            + address_gossiper
+            + config
+            + chainspec_loader
+            + storage
+            + contract_runtime
+            + linear_chain_fetcher
+            + linear_chain_sync
+            + block_validator
+            + deploy_fetcher
+            + block_executor
+            + consensus;
+
+        self.mem_total.set(total);
+        self.mem_metrics.set(metrics);
+        self.mem_network.set(network);
+        self.mem_small_network.set(small_network);
+        self.mem_address_gossiper.set(address_gossiper);
+        self.mem_config.set(config);
+        self.mem_chainspec_loader.set(chainspec_loader);
+        self.mem_storage.set(storage);
+        self.mem_contract_runtime.set(contract_runtime);
+        self.mem_linear_chain_fetcher.set(linear_chain_fetcher);
+        self.mem_linear_chain_sync.set(linear_chain_sync);
+        self.mem_block_validator.set(block_validator);
+        self.mem_deploy_fetcher.set(deploy_fetcher);
+        self.mem_block_executor.set(block_executor);
+        self.mem_consensus.set(consensus);
+
+        // Stop the timer explicitly, don't count logging.
+        let duration_s = timer.stop_and_record();
+
+        // Set the estimated flag to `true` as we have checked that the metrics were collected at
+        // least once
+
+        debug!(
+        %total,
+        %duration_s,
+        %metrics,
+        %network,
+        %small_network,
+        %address_gossiper,
+        %config ,
+        %chainspec_loader,
+        %storage ,
+        %contract_runtime,
+        %linear_chain_fetcher,
+        %linear_chain_sync,
+        %block_validator,
+        %deploy_fetcher,
+        %block_executor,
+        %consensus,
+        "Collected new set of memory metrics for the joiner");
+    }
+}
+
+impl Drop for MemoryMetrics {
+    fn drop(&mut self) {
+        self.registry
+            .unregister(Box::new(self.mem_total.clone()))
+            .expect("did not expect deregistering joiner_mem_total to fail");
```

Comment:
We shouldn't really run the risk of panicking during a drop, since that could trigger a panic while panicking.  These `expect` calls should probably just log any error.  If that's in the other `MemoryMetrics` for the other reactors, we should change them too.

Reply:
@marc-casperlabs you know about the wire up for metrics. Thoughts?

Reply:
Changed to `warn!` instead of a panic